### DISCUSSION TITLE
Extend the USB RNDIS (And ECM/NCM) example to the ESP32-S2 and ESP32-S3

### DISCUSSION THREAD

**Discussion by FunDeckHermit** [cesanta/mongoose#2506] [Feature requests]
I've been trying to get a working USB based http server up and running using ESP-IDF and LwIP on the ESP32-S3. Then I found the excellent RP2040 example and got it up and running in 5 minutes. 
It would be nice to have these examples for the ESP32-S2 and ESP32-S3 as they also support native USB.

**Comment by scaprile** [comment]
The RP2040 examples uses Mongoose built-in TCP/IP stack and TinyUSB
The ESP-IDF uses lwIP, and I don't know which USB stack it uses.
We wrote descriptors for TinyUSB, it already handles USB networking devices
We then wrote a driver for our own TCP/IP stack.
Depending on what the IDF uses, if having two different TCP/IP stacks is no problem, you can port the example. To be honest, I don't think we are interested in writing a driver for lwIP nor messing with the IDF or socket/USB stuff in it.

**Comment by FunDeckHermit** [comment]
Thank you for your reply!
I was under the impression that the LwIP is quite loosely coupled into the application as an ESP-IDF component. 
Seeing as Mongoose is also a [component ](https://components.espressif.com/components/cesanta/mongoose) in ESP-IDF, Mongoose could compete with LwIP on the ESP32 platform. Supporting LwIP examples doesn't make sense for Mongoose, that's certain. 

ESP-IDF also uses TinyUSB as their USB stack.

**Reply by scaprile** [reply]
For modules with no WiFi/Ethernet, using Mongoose TCP/IP stack is IMO the right choice. There might be some tweaks needed to build.
For modules with both USB and WiFi/Ethernet, doing that will mean that Mongoose will only work on the USB network interface, it won't see both and enabling its builtin driver disables the socket interface. In that case, to use both interfaces a TinyUSB netif for lwIP would be needed.
For modules with WiFi/Ethernet (no USB), Mongoose will use the socket interface and run over the ESP-IDF stack (modified lwIP)

**Comment by imdahisaria** [comment]
Hi All, I came across this post and this is something exactly i need. I started a discussion https://github.com/cesanta/mongoose/discussions/3597

Would appreciate if someone could point in the right direction how can i get this to work . Thank you!

**Reply by scaprile** [reply]
Please DO NOT necropost
